### PR TITLE
Wrote simple streaming dataloader

### PR DESCRIPTION
**Pull request by lennart-finke**
## Description
Added a streaming dataloader to avoid storing large datasets.

## Related Issue
Progress towards #8

### CODE REVIEW

**Comment by danbraunai-apollo** [comment]
Does this still work with distributed data parallel? I'm a bit concerned that this setup is more complex than needed. In general I think we'll need to support:
1. The dataset and tokenizer will be hosted on huggingface.
2. The pre-tokenized dataset will be hosted on huggingface (so we don't have to tokenize it on the fly everytime we train).

I think we can just get away with using huggingface's load_dataset with streaming=True. An example is [here](https://github.com/ApolloResearch/e2e_sae/blob/83b4add4d2652a4e5c6775a0e2b752897c871a87/e2e_sae/data.py#L134), which supports loading tokenized or untokenized datasets. Then we would just need to set it up to work for DDP. Not sure of the easiest way, there's probably standard setups here, maybe using a [distributed sampler](https://pytorch.org/docs/stable/data.html#torch.utils.data.distributed.DistributedSampler).

I've added the above to #8 (sorry I should have done this earlier).

**Comment by lennart-finke** [comment]
When writing it I was thinking about distributed data parallel, and it could work, but I have not been able to test that.

Currently the code assumes that the the tokenizer comes from tiktoken and that the streamed dataset is not tokenized.

Am open to closing the PR if there is an easier solution; that decision I would leave to you.

**Comment by lennart-finke** [comment]
After looking into this a bit more, there is also a builtin in Huggingface for this, the function [`split_dataset_by_node `](https://github.com/huggingface/datasets/issues/5360). If that make sense if I'll adapt the file you mentioned and use this next.

**Review by danbraunai — APPROVED** [review]
Looks good! Noting that I didn't run anything. There are no tests to check for things like the dataloader distributing things correctly and the tokenizer working properly. Adding such tests would be valuable, but also could be time consuming.

At a minimum, you should check that the dataloader does give you different data on each processes (when you start doing multiprocessing) and that the tokenizer works as you'd expect.

File: simple_stories_train/train_llama.py [hunk 1]
```diff
@@ -1019,7 +943,7 @@ def get_lr(it: int) -> float:
                 # addition of gradients corresponds to a SUM in the objective, but
                 # instead of a SUM we want MEAN, so we scale the loss here
                 loss = loss / grad_accum_steps
-                lossf += loss.detach()  # keep track of the mean loss
+                lossf += loss.item()  # this was .detach() before, and I don't know why
```
[the remark this answers is not in the document]
Comment (lennart-finke) [reply simple_stories_train/train_llama.py:946]: Thanks for the explanation, makes sense

File: simple_stories_train/train_llama.py [hunk 2]
```diff
@@ -949,15 +873,15 @@ def get_lr(it: int) -> float:
         last_step = step == args.num_iterations
 
         # once in a while evaluate the validation dataset
-        if (args.val_loss_every > 0 and (step % args.val_loss_every == 0 or last_step)) and (
-            val_loader is not None
-        ):
+        if (args.val_loss_every > 0 and (step % args.val_loss_every == 0 or last_step)):
             model.eval()
-            val_loader.reset()
+            # val_loader.reset() # What was reset good for again?
```
[the remark this answers is not in the document]
Comment (lennart-finke) [reply simple_stories_train/train_llama.py:878]: So `torch.util.data.DataLoader`, which we're now using as the dataloader class, seems to only have an internal `_reset()`. I hope it's fine to just re-instantiate the iterator over the DataLoader in that line, which I have now done as a fix?

File: simple_stories_train/tokenizer/stories-3072.json
[the remark this answers is not in the document]
Comment (lennart-finke) [reply simple_stories_train/tokenizer/stories-3072.json:1]: Yes! I hope it won't change all too often for now...

File: simple_stories_train/tokenizer.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+from tokenizers import Tokenizer, pre_tokenizers
+from tokenizers.models import WordPiece
+from tokenizers.pre_tokenizers import Digits, Punctuation, Whitespace
+from tokenizers.trainers import WordPieceTrainer
+
+OUT_DIR = Path("tokenizers")
+
+
+def clean_dataset():
+    dataset = load_dataset("lennart-finke/SimpleStories")
```
[the remark this answers is not in the document]
Comment (lennart-finke) [reply simple_stories_train/tokenizer.py:19]: Quickly parameterized the dataset name but yup, that was just copy&pasted wholesale

**Comment by lennart-finke** [comment]
> Looks good! Noting that I didn't run anything. There are no tests to check for things like the dataloader distributing things correctly and the tokenizer working properly. Adding such tests would be valuable, but also could be time consuming.

Good idea, I added a minimal test for the dataloader, in particular that it gives back the right shape and that different ddp ranks give different data.

> At a minimum, you should check that the dataloader does give you different data on each processes (when you start doing multiprocessing) and that the tokenizer works as you'd expect.

Right, the tokenizer is still a bit of a mystery to me, and the current behaviour is likely not what we want.

For instance, encoding and decoding a story

> In a magical forest, there was a kind bear who loved to help. One day, he found a sparkling gem among the leaves. Curiously, he touched it and turned into a fluffy bunny! The bunny hopped with joy but soon noticed a sad squirrel up in a tree. The bunny hopped closer and asked, "Why are you sad?" The squirrel replied, "I can't find my acorns. I need them for winter!" The bunny wanted to help, so he put on his bunny ears and searched everywhere in the forest. "Don't worry, I'll find your acorns!" he promised.

gives

> a magical forest , there was a kind bear who loved to help . day , he found a sparkling gem among the leaves . , he touched it and turned into a fluffy bunny ! bunny hopped with joy but soon noticed a sad squirrel up in a tree . bunny hopped closer and asked , " are you sad ? " squirrel replied , " can \' t find my acorn ##s . need them for winter ! " bunny wanted to help , so he put on his bunny ears and searched everywhere in the forest . " \' t worry , \' l ##l find your acorn ##s ! " he promised .

Now that you point it out like that, I should at least understand what's happening there before merging...

**Comment by lennart-finke** [comment]
Now I understand what was going on, the dataloader needs to convert to lowercase for this tokenizer. Now it works as expected!

By the way, any thought about whether to convert to lowercase and remove whitespace, as is currently done?

**Comment by danbraunai** [comment]
> By the way, any thoughts about whether to convert to lowercase and remove whitespace, as is currently done?
No opinions myself, I haven't been following the tokenizer stuff. @nix-apollo?

File: simple_stories_train/train_llama.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (danbraunai) [reply simple_stories_train/train_llama.py]: Ah OK yeah. The downside compared to doing reset with the DistributedDataLoader is that re-instantiating would load all of the data to memory again, which could be costly in time and memory (if it still has the old one in memory). If you play around and don't see an issue with memory or time bumps in either of those, then it's probably fine. Otherwise you might want to use DistributedDataLoader like the original script.

[the remark this answers is not in the document]
Comment (lennart-finke) [reply simple_stories_train/train_llama.py]: Alright, will keep it in mind!

**Comment by nix-apollo** [comment]
> > By the way, any thoughts about whether to convert to lowercase and remove whitespace, as is currently done?
> 
> No opinions myself, I haven't been following the tokenizer stuff. @nix-apollo?

My opinion is that the tokenizer lowercasing and removing whitespace is worth it for most research usecases. But it also adds complexity and confusion for people as evidenced here. Tokenizers not preserving all information when tokenizing and detokenizing could certainly lead to bugs in the user's code, for instance (although at least in this case it's pretty obvious).

Currently on balance I am in favor of lowercase and some form of whitespace normalizing, but that's a weakly held opinion.

**Comment by lennart-finke** [comment]
I see, thanks! Then I propose to just keep with the current tokenizer setup for the beta training run and maybe adjust for the final run if we encounter issues with it.
